src/overlays/actors/ovl_Obj_Grass_Unit/z_obj_grass_unit.c

```diff
@@ -5,34 +5,161 @@
+    PolarCoord* grassPos;
+    CollisionPoly* poly;
+    s32 bgId;
+    ObjGrassElement* grassElem;
+    ObjGrassUnitPattern* grassPattern;
+    s8 dropTable;
+
+    dropTable = ((s16)this->params >> 8) & 0x1F;
+    if ((sGrassManager == NULL) && !ObjGrassUnit_SpawnObjGrass(this, play)) {
+        Actor_Kill(this);
+        return;
+    }
+    if ((sGrassCarry0 == NULL) && ObjGrassUnit_SpawnObjGrassCarry0(this, play)) {
+        sGrassCarry0->grassManager = sGrassManager;
+    }
+    if ((sGrassCarry1 == NULL) && ObjGrassUnit_SpawnObjGrassCarry1(this, play)) {
+        sGrassCarry1->grassManager = sGrassManager;
+    }
+    if (!sInitialized && (sGrassManager != NULL) && (sGrassCarry0 != NULL) && (sGrassCarry1 != NULL)) {
+        sInitialized = true;
+        sGrassManager->grassCarry[0] = sGrassCarry0;
+        sGrassManager->grassCarry[1] = sGrassCarry1;
+        sGrassCarry0->grassManager = sGrassManager;
+        sGrassCarry1->grassManager = sGrassManager;
+    }
+
+    grassManager = sGrassManager;
+    if (sGrassManager->activeGrassGroups >= ARRAY_COUNT(sGrassManager->grassGroups)) {
+        Actor_Kill(this);
+        return;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Grass_Unit/func_809ABE54.s")
+    homePosYSum = 0.0f;
+    grassGroup = &grassManager->grassGroups[sGrassManager->activeGrassGroups];
+    grassGroup->count = 0;
+    grassPattern = &sGrassPatterns[(this->params & 1)];
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Grass_Unit/func_809ABEC4.s")
+    for (i = 0; i < grassPattern->count; i++) {
+        grassElem = &grassGroup->elements[grassGroup->count];
+        grassPos = &grassPattern->positions[i];
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Grass_Unit/func_809ABF38.s")
+        grassElem->pos.x = (Math_CosS((this->home.rot.y + grassPos->rotY)) * grassPos->distance) + this->home.pos.x;
+        grassElem->pos.y = this->home.pos.y + 100.0f;
+        grassElem->pos.z = (Math_SinS((this->home.rot.y + grassPos->rotY)) * grassPos->distance) + this->home.pos.z;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Obj_Grass_Unit/ObjGrassUnit_Init.s")
+        grassElem->pos.y = BgCheck_EntityRaycastFloor5(&play->colCtx, &poly, &bgId, this, &grassElem->pos);
+        tmp = grassElem->pos.y - this->home.pos.y;
+        if ((fabsf(tmp) < 80.0f) && (grassElem->pos.y > BGCHECK_Y_MIN)) {
+            grassGroup->count++;
+            grassElem->rotY = (s16)(Rand_Next() >> 0x10);
+            grassElem->dropTable = dropTable;
+            if (ObjGrassUnit_IsUnderwater(play, &grassElem->pos)) {
+                grassElem->elem_flags |= 8;
+            }
+            homePosYSum += grassElem->pos.y;
+        }
+    }
+    if (grassGroup->count > 0) {
+        grassManager->activeGrassGroups++;
+        grassGroup->homePos.x = this->home.pos.x;
+        grassGroup->homePos.y = (homePosYSum / grassGroup->count);
+        grassGroup->homePos.z = this->home.pos.z;
+    }
+    Actor_Kill(this);
```

Comment:
this actor kills itself after init?

Reply:
Yep. The setup is pretty weird and nuanced, so perhaps I should put in a comment explaining how things work. 

You have Obj_Grass, which is the manager actor. It's purpose is to reduce the cost of having a lot of grass spawned at once by only processing a maximum 20 grass bushes on any given update cycle. 

Obj_Grass_Unit is the client of Obj_Grass. It initializes Obj_Grass if it hasn't been already, and then passes the spawn information over to the manager. Once this happens, there's nothing else for it to do so it just dies.

The nuanced thing here is scope. `Obj_Grass_Unit` is responsible for initializing and holding a reference to the Obj_Grass instance. But because ovl_Obj_Grass_Unit uses `ALLOCTYPE_NORMAL` allocation, the manager reference will fall out of scope once all instances are killed. In other words, if you spawn a "burst" of `Obj_Grass_Unit` in one frame (e.g. loading a new room), those `Obj_Grass_Unit` will use the same Obj_Grass instance, but subsequent spawns a few frames later will initialize a completely new manager.